README.md
# Mini File Manager

Mini File Manager is a file management interface for your Symfony backend. The connection is made with [pentatrion/upload-bundle](https://github.com/lhapaipai/upload-bundle).

Mini File Manager is very flexible, you can :
  - upload big files / resume uploads with [resumableJS](https://github.com/23/resumable.js/) integration.
  - crop / resize / rotate image after their upload.
  - custom easily the theme with CSS variables.
  - localize the application.

You can see an example here : [**Live Demo**](https://mini-file-manager.pentatrion.com/) (TinyMCE integration / File Picker with form)

<img alt="Mini File Manager" src="https://user-images.githubusercontent.com/1088155/128615403-2b41fbb4-dd4e-452e-b2c2-6926642bf146.jpg">

## Getting started

if you want a practical [**minimal installation recipe**](https://github.com/lhapaipai/mini-file-manager/blob/main/docs/getting-started.md) you can follow this tutorial which will not take you more than 5 minutes.

if you want a full example you can consult the sources of [mini-file-manager-template](https://github.com/lhapaipai/mini-file-manager-template). This example contain full integration of the manager with form (file picker, wysiwyg), and simple file manager.

## Dependances

Mini File Manager is designed with Vue (unless you modify the sources, the dependencies are integrated into the module) and **you have to configure a Symfony backend to manage your files**.

i18n integration, available in English and French or provide your own Translation file. PRs are welcome.

<p align="center">
  <img width="100" src="https://raw.githubusercontent.com/lhapaipai/mini-file-manager/main/docs/symfony.svg" alt="Symfony logo">
  <img width="100" src="https://raw.githubusercontent.com/lhapaipai/mini-file-manager/main/docs/vue.svg" alt="Vue logo">
</p>

- Symfony v5
  - pentatrion/upload-bundle
  - liip/imagine-bundle
  - symfony/validator


## Installation

### BackEnd

before you configure the mini-file-manager, you must first install and configure the backend under Symfony with [pentatrion/upload-bundle](https://github.com/lhapaipai/upload-bundle).

### FrontEnd

```console
npm install mini-file-manager
```

Copy `dist/file-manager` directory from `node_modules/mini-file-manager` to your webroot directory. it contains icons for each file type.

Copy `dist/fonts` directory from `node_modules/mini-file-manager` to your webroot directory. it contains icons for buttons.

#### with UMD / ES Modules

`mini-file-manager.umd.js` provides `miniFileManager` global variable.

```html
<!-- with UMD -->
<link rel="stylesheet" href="/dist/mini-file-manager.css" />
<div id="file-manager"></div>
<script src="/dist/mini-file-manager.umd.js"></script>
<script>
  let {
    FileManager,
    FileManagerModal,
    FormFilePicker
  } = miniFileManager;
    // etc...
</script>
```

or

```js
// with bundler
import {
  FileManager,
  FileManagerModal,
  FormFilePicker
} from "mini-file-manager";
import "mini-file-manager/dist/mini-file-manager.css";
```

## Configuration

Mini File Manager export 3 functions

```js
FileManager(
  "#selector",        // string | HTMLElement
  fileManagerOptions  // FileManagerOptions
);

FileManagerModal(
  fileManagerOptions, // FileManagerOptions
  onSuccess,          // (selectedFiles: Files[]) => void
  onAbort             // () => void
);

FormFilePicker(
  inputElt,             // string | HTMLElement
  formFilePickerOptions,// FormFilePickerOptions
  fileManagerOptions,   // FileManagerOptions
  selection             // Files[]
)
```

```ts
const options: FileManagerOptions = {
  endPoint: "http://url-to-backend.com/media-manager",
  entryPoints: [
    {
      label: "Conversation",
      // base directory relative to origin
      directory: "projet/mon-projet",
      origin: "private_uploads",
      // readOnly mode -> can't upload/modify/rename/delete files
      readOnly: false,
      icon: "fa-lock",
    },
  ],

  // if you wants to filter files you can select
  // only for the modal "openFileManager"
  fileValidation: {
    mimeGroup: "image",
    allowDir: false,
    imageOptions: {
      allowSvg: false,
      width: 1200,
      height: 900,
      minWidth: 1200,
      maxWidth: 800,
      minHeight: 1200,
      maxHeight: 800,
      ratio: 0.66, // float number : width/height

      // note : if you give a width and a height, the ratio is calculated
      // and only the width and the ratio are used.
    },
  }

  // if you wants to filter files you can upload
  fileUpload: {
    maxFileSize: 500 * 1024 * 1024, // 500Mo
    fileType: [
      "text/*",
      "image/*", // image/vnd.adobe.photoshop  image/x-xcf
      "video/*",
      "audio/*",
      "application/rtf",
      "application/pdf",
      "application/xml",
      "application/zip",
      "font/ttf",
      "font/woff",
      "font/woff2",
      "application/vnd.oasis.opendocument.spreadsheet", // libreoffice ods
      "application/vnd.oasis.opendocument.text", // libreoffice odt
      "application/vnd.openxmlformats-officedocument.wordprocessingml.document", // docx
      "application/msword", // doc
      "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet", // xlsx
      "application/vnd.ms-excel", // xls
      "application/json",
      "application/illustrator", // .ai
    ]
  },
  
  locale: "en", // "en" | "fr" | "custom"
  localeData: {
    // if "custom" write here your translations by referring to the file
    // https://github.com/lhapaipai/mini-file-manager/blob/main/src/locales.js
    apply: "foo",
    cancel: "bar",
    editAndSelect: "baz",
    // ...
  }

  multiple: false,

  // not needed with FormFilePicker
  // selection is retrieved from input value
  originalSelection: ["posts/autre/ign.jpg"],

  theme: "pentatrion-theme"
};

const formFilePickerOptions: FormFilePickerOptions = {
  type: "image",   // "image" | "file"
  filter: "small", // any filter defined in LiipImagineBundle
  multiple: true,
};
```

### Theme

if you wants to define your own theme, you can set the `theme` option with custom class and redefine the css variables from `src/css/variables.css` with this class.

```js
const options = {
  //...
  theme: 'my-own-theme'
};
```

```css
.my-own-theme {
  --primary-color-light: #fff9d2;
  --primary-color: #ffe64b;
  --primary-color-active: #fadf30;
  --primary-color-dark: #eac800;

  /* etc... */
}
```


## Screenshots

<img alt="Crop image" src="https://user-images.githubusercontent.com/1088155/128615409-8ba709cf-dd51-40c8-a5f5-93aaacd98fe3.jpg"><br>

### Mobile friendly

<img alt="Mobile first" style="border: 1px solid #aaa; margin-right: 10px;" src="https://user-images.githubusercontent.com/1088155/128615401-10dca575-3beb-4cc2-885f-d80498d181d6.jpg" width="400"><img alt="Crop mobile first" style="border: 1px solid #aaa;" src="https://user-images.githubusercontent.com/1088155/128623126-8fdda390-fc3a-455e-a8ad-5f3ab9d7d2e1.jpg" width="400">


CHANGELOG.md
v1.0.9

- fix bug with slash when entrypoint directory is empty string
- fix bug with FormFilePicker when selecting the same file with `multiple: false`.

src/utils.js
export function parseOriginalSelection(originalSelection, entryPoints) {
  if (!originalSelection || originalSelection.length === 0) {
    return null;
  }

  let newList = [],
    temp,
    origin,
    path,
    originPos,
    dirPos,
    dir,
    id;
  for (let fileId of originalSelection) {
    if (fileId[0] !== "@") {
      origin = entryPoints[0].origin;
      path = fileId;
      id = `@${origin}:${path}`;
    } else {
      id = fileId;
      temp = fileId.substr(1);
      originPos = temp.indexOf(":");
      origin = temp.substr(0, originPos);
      path = temp.substr(originPos + 1);
    }
    dirPos = path.lastIndexOf("/");
    if (dirPos === -1) {
      dir = "";
    } else {
      dir = path.substr(0, dirPos);
      path = path.substr(dirPos + 1);
    }
    newList.push([origin, dir, path, id]);
  }
  return newList;
}

export function verifyValidation(fileValidation) {
  if (!fileValidation) return null;
  if (fileValidation.imageOptions) {
    if (fileValidation.imageOptions.width && fileValidation.imageOptions.height) {
      fileValidation.imageOptions.ratio =
        fileValidation.imageOptions.width / fileValidation.imageOptions.height;
      delete fileValidation.imageOptions.height;
    }
  }
  return fileValidation;
}

export function verifyUploadOptions(fileUpload) {
  return Object.assign(
    {
      maxFileSize: 500 * 1024 * 1024,
      fileType: [
        "text/*",
        "image/*", // image/vnd.adobe.photoshop  image/x-xcf
        "video/*",
        "audio/*",
        "application/rtf",
        "application/pdf",
        "application/xml",
        "application/zip",
        "font/ttf",
        "font/woff",
        "font/woff2",
        "application/vnd.oasis.opendocument.spreadsheet", // tableur libreoffice ods
        "application/vnd.oasis.opendocument.text", // traitement de texte odt
        "application/vnd.openxmlformats-officedocument.wordprocessingml.document", // docx
        "application/msword", // doc
        "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet", // xlsx
        "application/vnd.ms-excel", // xls
        "application/json",
        "application/illustrator", // .ai
      ],
    },
    fileUpload,
  );
}

export function urlizer(value, separator = "-") {
  let src = "áàãâéèêëÿíïóôõúüñçÁÀÃÂÉÊÈÍÓÔÕÚÜÑÇ";
  let dst = "aaaaeeeeyiiooouuncAAAAEEEIOOOUUNC";
  let map = {};
  for (let i = 0; i < src.length; i++) {
    map[src.charAt(i)] = dst.charAt(i);
  }
  map["æ"] = "ae";
  map["œ"] = "oe";
  map[" "] = separator;

  function replaceAll(str, find, replace) {
    return str.replace(new RegExp(find, "g"), replace);
  }

  value = value.toLowerCase();
  value = replaceAll(value, "-", " ");
  value = replaceAll(value, "_", " ");
  let tmp = "";
  let splited = value.split(" ");
  for (let i = 0; i < splited.length; ++i) {
    let s = splited[i];
    if (s.trim().length != 0) {
      tmp += " " + s.trim();
    }
  }
  value = tmp.trim();
  let out = "";
  for (let i = 0; i < value.length; ++i) {
    let char = value.charAt(i);
    let translatedChar = map[char];

    if (translatedChar != null) {
      out += translatedChar;
    } else if ((char >= "a" && char <= "z") || (char >= "0" && char <= "9")) {
      out += char;
    }
  }
  return out;
}

export function sanitizeFilename(filename, filenameList) {
  let filenameWithoutExtension,
    extension = "",
    extPos = filename.lastIndexOf(".");
  if (extPos !== -1) {
    extension = filename.substr(extPos).toLowerCase();
    filenameWithoutExtension = filename.substr(0, extPos);
  } else {
    filenameWithoutExtension = filename;
  }

  filenameWithoutExtension = urlizer(filenameWithoutExtension);

  if (filenameList.find((f) => f.filename === filenameWithoutExtension + extension)) {
    let counter = 1;
    while (
      filenameList.find(
        (f) => f.filename === `${filenameWithoutExtension}-${counter}${extension}`,
      )
    ) {
      counter++;
    }
    filenameWithoutExtension = `${filenameWithoutExtension}-${counter}`;
  }

  return filenameWithoutExtension + extension;
}

export function filename2dirname(filename) {
  let extPos = filename.lastIndexOf(".");
  if (extPos !== -1) {
    return filename.substr(0, extPos) + filename.substr(extPos + 1);
  } else {
    return filename;
  }
}

export function isValidFile(file, fileValidation) {
  if (!fileValidation) {
    return true;
  }
  if (fileValidation.allowDir === false && file.type === "dir") {
    return false;
  }

  if (fileValidation.mimeGroup && fileValidation.mimeGroup !== file.mimeGroup) {
    return false;
  }

  if (fileValidation.mimeGroup === "image") {
    if (!fileValidation.imageOptions) {
      return true;
    }
    if (
      fileValidation.imageOptions.allowSvg &&
      (file.mimeType === "image/svg+xml" || file.mimeType === "image/svg")
    ) {
      return true;
    }

    // si on a des critères de dimension et que l'image n'en dispose pas
    // comme du svg alors le fichier n'est pas valide.
    if (!file.details) {
      return false;
    }

    if (
      fileValidation.imageOptions.ratio &&
      Math.round(fileValidation.imageOptions.ratio * 100000) !==
        Math.round(file.details.ratio * 100000)
    ) {
      return false;
    }

    if (
      fileValidation.imageOptions.width &&
      fileValidation.imageOptions.width !== file.details.width
    ) {
      return false;
    }
    if (
      fileValidation.imageOptions.height &&
      fileValidation.imageOptions.height !== file.details.height
    ) {
      return false;
    }

    if (
      fileValidation.imageOptions.minWidth &&
      fileValidation.imageOptions.minWidth > file.details.width
    ) {
      return false;
    }
    if (
      fileValidation.imageOptions.minHeight &&
      fileValidation.imageOptions.minHeight > file.details.height
    ) {
      return false;
    }

    if (
      fileValidation.imageOptions.maxWidth &&
      fileValidation.imageOptions.maxWidth < file.details.width
    ) {
      return false;
    }
    if (
      fileValidation.imageOptions.maxHeight &&
      fileValidation.imageOptions.maxHeight < file.details.height
    ) {
      return false;
    }
  }

  return true;
}

export function isEditableFile(file) {
  if (file.mimeGroup !== "image") {
    return false;
  }
  if (file.mimeType === "image/svg+xml" || file.mimeType === "image/svg") {
    return false;
  }
  return true;
}

export async function createFileInfosFromUpload(fileUploadInfos, directory, origin) {
  let uploadRelativePath = `${directory}/${fileUploadInfos.fileName}`;
  let mimeType = fileUploadInfos.file.type;
  let mimeGroup = mimeType.split("/")[0];

  let infos = {
    inode: null,
    id: fileUploadInfos.id,
    filename: fileUploadInfos.fileName,
    directory,
    uploadRelativePath,
    mimeType,
    mimeGroup,
    type: "file",
    uploader: "",
    origin,
    size: fileUploadInfos.file.size,
    humanSize: humanFileSize(fileUploadInfos.file.size),
    createdAt: new Date().toISOString(),
    isDir: false,
    url: null,
    urlTimestamped: null,
    icon: "/file-manager/icons/other.svg",
    uploadInfos: {
      progression: 0,
      retry: 0,
    },
  };

  if (mimeGroup === "image") {
    infos.thumbnails = {
      small: await generateThumbnail(fileUploadInfos.file, [250, 250]),
    };
  }

  return infos;
}

export function generateThumbnail(file, boundBox) {
  if (!boundBox || boundBox.length != 2) {
    throw "You need to give the boundBox";
  }
  // let scaleRatio = Math.min(...boundBox) / Math.max(file.width, file.height);
  let reader = new FileReader();
  let canvas = document.createElement("canvas");
  let ctx = canvas.getContext("2d");

  return new Promise((resolve) => {
    reader.onload = function (event) {
      let img = new Image();
      img.onload = function () {
        let scaleRatio = Math.min(...boundBox) / Math.max(img.width, img.height);
        let w = img.width * scaleRatio;
        let h = img.height * scaleRatio;
        canvas.width = w;
        canvas.height = h;
        ctx.drawImage(img, 0, 0, w, h);
        return resolve(canvas.toDataURL(file.type));
      };
      img.src = event.target.result;
    };
    reader.readAsDataURL(file);
  });
}

export function humanFileSize(size) {
  let i = Math.floor(Math.log(size) / Math.log(1024));
  return (
    (size / Math.pow(1024, i)).toFixed(2) * 1 + " " + ["B", "Ko", "Mo", "Go", "To"][i]
  );
}
